### ISSUE TITLE
Avoid unnecessary reloading of the document

### ISSUE THREAD

**Issue by TheDcoder** [asciidoctor/asciidoctor-browser-extension#315]
I see that the extension attempts to reload the document on numerous occasions where reloading isn't warranted. I have the auto-reload setting disabled for both remote and local files, but even then the extension attempts to reload the document on several occasions, here are the two occasions which are prominent:

1. On clicking a section from ToC
2. Activating the tab after a bit of time

I do not see a reason why the document needs to be reloaded in these cases. Normally I would have just ignored this sort of bug if it wasn't causing any disruption, but in this case it triggers another bug which totally wipes the document.

After updating Chromium to version 80, it has started blocking requests to local files due to CORS:
```
Access to XMLHttpRequest at 'file:///tmp/foo.adoc?_=1581403964282' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, https.
```
(This can be easily reproduced by disabling and enabling rendering of the document, which forces the extension to refetch the source text for some reason)

So TL;DR, due to Chrome's new CORS policy, whenever the source text is reloaded via a `XMLHttpRequest`, it is blocked and therefore rendering a blank page, and this happens frequently even with auto-reload turned off.

Please fix it as soon as possible.

**Comment by Mogztter** [comment]
Hello @TheDcoder 

If you've disabled auto-reload it should not reload. I don't know if it's a change in Chrome 80 but I cannot reproduce this behavior in Chrome Version 79.0.3945.79 (Linux):

![no-reload](https://user-images.githubusercontent.com/333276/74218730-0b3eb900-4cab-11ea-8f11-daaabdac322e.gif)

Am I missing something?

**Comment by TheDcoder** [comment]
Hello @Mogztter,

> I don't know if it's a change in Chrome 80 but I cannot reproduce this behavior in Chrome Version 79.0.3945.79 (Linux)

The issue probably doesn't exist in 79 because it only started happening after I updated to 80. Also, the most effective way to reproduce the CORS block is to enable and disable rendering by clicking the extension's icon in the toolbar. This reproduces the error with a 100% chance for me, the other two scenarios I mentioned have less than 100% chance of occuring, I guess they depend on a certain condition which does not happen always.

> Am I missing something?

Maybe the latest version of Chrome :smile:

I also want to mention more technical details that I managed to extract by debugging the content script, it looks like the background script is sending the `extension-enabled` status message sometimes when I activate the tab/window. This may give you some clue as to why it is reloading the document even with the auto-reload setting disabled.

**Comment by TheDcoder** [comment]
This literally happened while I was browsing through an already loaded document:
![image](https://user-images.githubusercontent.com/6347576/74228339-43172200-4ce6-11ea-9ec2-2fd0c206ebc6.png)

The extension recieves a `{"status":"extension-enabled"}` message out of the blue. The result after I resume the debugger is a blank page and an error about CORS in the console:
![image](https://user-images.githubusercontent.com/6347576/74228642-de0ffc00-4ce6-11ea-9122-d53f8aecdda7.png)

**Comment by Mogztter** [comment]
> The issue probably doesn't exist in 79 because it only started happening after I updated to 80.

Ok so it seems to be a regression in Chrome 80.
I found this changelog about XMLHTTPRequest: https://developers.google.com/web/updates/2019/12/chrome-80-deps-rems#disallow_synchronous_xmlhttprequest_in_page_dismissal

Could you please try to set `allow-sync-xhr-in-page-dismissal` as described in the link?

> Also, the most effective way to reproduce the CORS block is to enable and disable rendering by clicking the extension's icon in the toolbar. 

I'm using this technique to avoid reloading the tab/page (ie. https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/reload) and cache the raw content.

When you disable the extension, an XMLHttpRequest is sent to fetch the raw content (and cache it). When you enable the extension again, the raw content (cached) is used so it shouldn't use an XMLHttpRequest. But if the first XMLHttpRequest is blocked then the cache is empty and in this case we use another XMLHttpRequest to fetch the content.

We are also using XMLHttpRequest to fetch the content of included files.

> This literally happened while I was browsing through an already loaded document:

That's really strange. I believe that the event system is asynchronous but the event should only be sent when you click on the extension icon (to enable/disable the extension).

I will install Chrome 80 to do some tests.

**Comment by TheDcoder** [comment]
> Ok so it seems to be a regression in Chrome 80.
> I found this changelog about XMLHTTPRequest: https://developers.google.com/web/updates/2019/12/chrome-80-deps-rems#disallow_synchronous_xmlhttprequest_in_page_dismissal
> 
> Could you please try to set allow-sync-xhr-in-page-dismissal as described in the link?

Are you sure that this is related? I enabled the flag by going to `chrome://flags/#allow-sync-xhr-in-page-dismissal`, but I can still reproduce the issue. I do not think this is related to synchronous requests.

> That's really strange. I believe that the event system is asynchronous but the event should only be sent when you click on the extension icon (to enable/disable the extension).

I did not click on the extension icon when the event was sent, so I guess there is some other condition which triggers the event in the background script.

> I will install Chrome 80 to do some tests

That should shed more light onto why the event is being sent :)

**Comment by Mogztter** [comment]
> Are you sure that this is related? 

No, sorry about that.
The correct link describing the issue is: https://www.chromium.org/Home/chromium-security/extension-content-script-fetches/

As a workaround you can disable "web security", see https://github.com/asciidoctor/asciidoctor-browser-extension/issues/316#issuecomment-584553923

**Comment by Mogztter** [comment]
fixed in https://github.com/asciidoctor/asciidoctor-browser-extension/commit/4c20ea05f4526fe8b43bf142585d9606e83f417b
